File: go/274_H-Index.go
func hIndex(citations []int) int {
    var citationsCount = len(citations)
    var buckets = make([]int, citationsCount + 1)
    for i := range buckets {
        buckets[i] = 0          
    } 

    for _, citation := range citations  {
        if (citation >= citationsCount) {
            buckets[citationsCount]++
        } else {
            buckets[citation]++
        }
    }

    var count = 0;
    for i := citationsCount; i >= 0; i-- {
        count += buckets[i]
        if count >= i {
            return i
        }
    }

    return 0
}

File: php/274_H-Index.php
<?php

class Solution 
{

    /**
     * @param Integer[] $citations
     * @return Integer
     */
    function hIndex($citations) 
    {
        $citationsCount = count($citations);
        $buckets = array_fill(0, $citationsCount + 1, 0);

        foreach ($citations as $citation) {
            if ($citation >= $citationsCount) {
                $buckets[$citationsCount]++;
            } else {
                $buckets[$citation]++;
            }
        }

        $count = 0;
        for ($i = $citationsCount; $i >= 0; $i--) {
            $count += $buckets[$i];
            if ($count >= $i) {
                return $i;
            }
        }

        return 0;
    }
}

